### PR TITLE
ddl: check column and partition value have the same type for range column partition

### PR DESCRIPTION
**Pull request by tiancaiamao**
<!--
Thank you for contributing to TiDB! Please read TiDB's [CONTRIBUTING](https://github.com/pingcap/community/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

### What problem does this PR solve? <!--add issue link with summary if exists-->

MySQL:

```
mysql> create table t (
    ->   col datetime not null default '2000-01-01'
    -> ) partition by range columns (col) (
    ->   PARTITION p0 VALUES LESS THAN (20190905),
    ->   PARTITION p1 VALUES LESS THAN (20190906));
ERROR 1654 (HY000): Partition column values of incorrect type

mysql> create table t (
    ->   col datetime not null default '2000-01-01'
    -> ) partition by range columns (col) (
    ->   PARTITION p0 VALUES LESS THAN ('20190905'),
    ->   PARTITION p1 VALUES LESS THAN ('20190906'));
Query OK, 0 rows affected (0.02 sec)

mysql> alter table t add partition (partition p2 values less than (20190907));
ERROR 1654 (HY000): Partition column values of incorrect type
```

TiDB:

```
mysql> create table t (
    ->   col datetime not null default '2000-01-01'
    -> ) partition by range columns (col) (
    ->   PARTITION p0 VALUES LESS THAN (20190905),
    ->   PARTITION p1 VALUES LESS THAN (20190906));
Query OK, 0 rows affected (0.02 sec)
```

TiDB:
```
mysql> create table t (
    ->   col datetime not null default '2000-01-01'
    -> ) partition by range columns (col) (
    ->   PARTITION p0 VALUES LESS THAN ('20190905'),
    ->   PARTITION p1 VALUES LESS THAN ('20190906'));
Query OK, 0 rows affected (0.02 sec)

mysql> 
mysql> alter table t add partition (partition p2 values less than (20190907));
Query OK, 0 rows affected (0.01 sec)

mysql> show create table t;
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                                                                      |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `col` datetime NOT NULL DEFAULT '2000-01-01 00:00:00'
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY RANGE COLUMNS(col) (
  PARTITION p0 VALUES LESS THAN ("20190905"),
  PARTITION p1 VALUES LESS THAN ("20190906"),
  PARTITION p2 VALUES LESS THAN (20190907)
) |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### What is changed and how it works?

Fix the create table check for range column partition

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

 - Unit test


Related changes

 - Need to cherry-pick to the release branch


Release note

 - Write release note for bug-fix or new feature.

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @crazycs520 @zimulala @jackysp

**Comment by tiancaiamao** [comment]
PTAL @crazycs520 @zimulala

**Comment by tiancaiamao** [comment]
PTAL @crazycs520 @zimulala

**Comment by tiancaiamao** [comment]
PTAL @crazycs520 @zimulala

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -3554,6 +3563,12 @@ func buildPartitionInfo(meta *model.TableInfo, d *ddl, spec *ast.AlterTableSpec)
 		}
 		// For RANGE partition only VALUES LESS THAN should be possible.
 		clause := def.Clause.(*ast.PartitionDefinitionClauseLessThan)
+		if len(part.Columns) > 0 {
+			if err := checkRangeColumnsTypeAndValuesMatch(ctx, meta, part.Columns, clause.Exprs); err != nil {
```
Comment (zimulala) [inline ddl/ddl_api.go]: Could we put it to `checkPartitionByRange`? Then we can remove the `if s != nil ` in line 1621

Comment (tiancaiamao) [reply ddl/ddl_api.go]: Partition by range do not check value and column type match like partition by range columns.

Comment (tiancaiamao) [reply ddl/ddl_api.go]: Partition by range checks value expression evaluates to an integer...

Comment (zimulala) [reply ddl/ddl_api.go]: Got it.

**Pull request merged**